File: Python/Knapsack.py
def sort(l, m):
    for j in range(1, len(m)):
        for k in range(1, len(m)):
            if m[k] < m[k - 1]:
                tmp_m = m[k]
                m[k] = m[k - 1]
                m[k - 1] = tmp_m

                tmp_l = l[k]
                l[k] = l[k - 1]
                l[k - 1] = tmp_l
    return l, w


def knapsack(N, values, weights, max_weight):
    table = [[0 for x in range(max_weight + 1)] for y in range(N)]  # table[N][max_weight]

    sorted_values, sorted_weights = sort(values, weights)

    for i in range(N):
        for j in range(max_weight + 1):
            tangent_value = 0
            if sorted_weights[i] <= j:
                tangent_value = sorted_values[i] + table[i-1][j - sorted_weights[i]]

            table[i][j] = max(table[i - 1][j], tangent_value)

    print "Max value =", table[N - 1][max_weight]


v = [3, 7, 2, 9]
w = [2, 3, 4, 5]
knapsack(len(v), v, w, 5)

File: Java/src/Knapsack.java
/**
 * Created by Jay on 17/08/2016.
 */
public class Knapsack {
	private int N;
	private int maxWeight;
	private int values[];
	private int weights[];
	private int table[][];

	public Knapsack(int maxWeight, int values[], int weights[]) {
		this.maxWeight = maxWeight;
		this.values = values;
		this.weights = weights;
		N = weights.length;

		table = new int[N][maxWeight + 1];
		// to make sure...
		for (int i = 0; i < N; i++) {
			for (int j = 0; j <= maxWeight; j++)
				table[i][j] = 0;
		}
	}

	private void sortItems() {
		for (int j = 1; j < N; j++) {
			for (int i = 1; i < N; i++) {
				if (weights[i] < weights[i - 1]) {
					int tmpWeight = weights[i];
					weights[i] = weights[i - 1];
					weights[i - 1] = tmpWeight;

					int tmpValue = values[i];
					values[i] = values[i - 1];
					values[i - 1] = tmpValue;
				}
			}
		}
	}

	public void knapsack() {
		sortItems();
		for (int i = 0; i < N; i++) {
			for (int j = 0; j <= maxWeight; j++) {
				int above = 0;
				int tangent = 0;
				if (i > 0) {
					above = table[i - 1][j];
				}
				if (weights[i] <= j) {
					tangent = values[i];
					if (i > 0)
						tangent += table[i - 1][j - weights[i]];
				}
//				if (weights[i] <= j) {
//					tangent = values[i];
//					if (i > 0) {
//						above = table[i - 1][j];
//						tangent += table[i - 1][j - weights[i]];
//					}
//				}
				table[i][j] = max(above, tangent);
			}
		}
		System.out.println("Max value = " + table[N - 1][maxWeight]);
	}

	private int max(int x, int y) {
		if (x > y)
			return x;
		else
			return y;
	}

	public static void main(String[] args) {
		int values[] = {7, 3, 8, 6, 5};
		int weights[] = {10, 15, 2, 4, 5};

		new Knapsack(20, values, weights).knapsack();
	}

}
